_posts/2024-11-25-cheri-myths-safety-critical.markdown

```diff
@@ -0,0 +1,189 @@
+
+In the undefined-behaviour case, this is permitted to do absolutely anything.
+On conventional embedded systems, this may even include writing to a memory-mapped device register, so a single out-of-bounds write somewhere completely unrelated to device control may move an actuator connected to the physical part of a cyber-physical system.
+
+This is why memory-safety bugs are so bad: they can do *anything*.
+
+## What about MISRA C?
+
+MISRA C rule 1.3 says that a program may not contain any occurrences of undefined behaviour.
+As such, any program that is compliant with MISRA C will never trigger a CHERI trap.
+That's somewhat facetious, but it's mostly true.
+MISRA C doesn't stop there, it provides a load of other rules that help to avoid these problems.
+
+MISRA C is a fairly restrictive subset of C that tries to confine programmers to features that are amenable to static analysis (as well as avoiding other things that cause common bugs).
+The caveat with MISRA C is the same as that for Rust.
+Just as Rust allows you to write `unsafe`, MISRA C codebases allow you to document exemptions from rules.
+We found an instance of this in the FreeRTOS TCP/IP stack where a warning about rule 17.4 (return values must be used) was explicitly silenced.
+This was in an early initialisation function that would return an error code if allocation failed two calls lower down, which resulted in a global not being initialised.
+In normal use, this always succeeded because it happened during early boot, but when we introduced network stack restart then it could happen later and might fail.
+
+For the most part, MISRA C code should not trap with CHERI, and if it does then you probably haven't actually followed the MISRA C requirements.
+
+The same is true of Rust.
+If you're using purely safe Rust, your code will never trap on a CHERI system.
+If you're using `unsafe` in small, locally scoped places, you can provide guarantees that are not visible in the source code to ensure the same properties.
+If you write your entire program in unsafe Rust (no one does this!) then you should expect traps.
+
+So do I need CHERI if I'm using MISRA C or Rust?
+[I've answered that before](/cheri/myths/2024/08/28/cheri-myths-safe-languages.html), but broadly yes.
+CHERI lets you enforce properties even in the unsafe dialects of safe languages, in C and assembly code linked with them, and on binary-only components or components compiled with an untrusted compiler (or code that may be written specifically to exploit bugs in a trusted compiler).
+
+## What is your baseline?
+
+To answer the complaint about CHERI trapping, it's worth considering what happens if you have a memory-safety bug today.
+You will see one of the following:
+
+ - It's deterministically benign, reads some constant or write over something that's never read.
+ - It's deterministically out of a mapped / accessible region will be caught by your MMU / MPU and trap (so we are not worse)
```

Comment:
- It's deterministically outside of a mapped / accessible region so will be caught by your MMU / MPU and trap (therefore CHERI is no worse)